### PR TITLE
Add CQL event for tablet change

### PR DESCRIPTION
**Pull request by sylwiaszunejko**
The protocol extended with a new kind of event, cql_transport::event::event_type::TABLET_CHANGE.

It is sent whenever tablet metadata changes. It does not have information about the type of change
that happened just the information that something have changed.

I added the information about new event to the docs.

### CODE REVIEW

**Comment by sylwiaszunejko** [comment]
> I suspect most events do not have tests, which is unfortunate, but I'm hoping we won't keep this trend - can we add a test?

 Yes, events do not have test, and I do not have good idea where to add them/how they should look like.

**Comment by avikivity** [comment]
There are tablet tests, test/topology_experimental_raft/test_tablets.py

File: docs/dev/topology-over-raft.md [hunk 1]
```diff
@@ -164,6 +164,8 @@ When tablet is not in transition, the following invariants hold:
 1. The storage layer (database) on any node contains writes for keys which belong to the tablet only if
     that shard is one of the current tablet replicas.
 
+When tablet table changes there is an event notification sent (cql_transport::event::event_type::TABLET_CHANGE). Currently it only informs that something changed without specifying details. The body of the message is empty (except the event type).
+
```
Comment (avikivity) [inline docs/dev/topology-over-raft.md:168]: This needs to be documented in docs/dev/protocol-extensions.md.

We need to think about providing change information. The tablets table is liable to grow to 100k rows, we don't want all clients to reread 100k rows on every migration.

We might encode a list of rows with each event, each row corresponding to a row in the tablets table, and add special events for merge/split.

Comment (sylwiaszunejko) [reply docs/dev/topology-over-raft.md]: I moved this addition to documentation to `docs/dev/protocol-extensions.md`.

**Review by piodul — COMMENTED** [review]
How does the driver know whether it can subscribe to the new event? The older versions of scylla will not recognize it. How is it going to be handled?

We already have a framework for negotiation of protocol features, maybe we can use it (the LWT and rate limit error extensions already do).

**Comment by sylwiaszunejko** [comment]
> How does the driver know whether it can subscribe to the new event? The older versions of scylla will not recognize it. How is it going to be handled?
> 
> We already have a framework for negotiation of protocol features, maybe we can use it (the LWT and rate limit error extensions already do).

If I understand correctly for driver to receive new types of events it has to register for it, so if there is no change in the driver my change should not break anything, due to no registration. On the other hand if there will be change in the driver, and if it use old version of Scylla it will register for new type of event, but no event with this new type will be received, but it would not cause an error, am I right (I tested it only on python-driver)?

**Comment by avikivity** [comment]
We can document that if the tablets table is available, then the event is available too.

**Comment by avikivity** [comment]
Another option is to return an ERROR for the REGISTER message. Errors aren't documented for REGISTER, so we have to check if Cassandra returns an error for an unrecognized event, or if it swallows it silently.

**Comment by sylwiaszunejko** [comment]
> Another option is to return an ERROR for the REGISTER message. Errors aren't documented for REGISTER, so we have to check if Cassandra returns an error for an unrecognized event, or if it swallows it silently.

I checked that and it depends on the driver, python-driver doesn't returns error, everything works fine, but gocql for example returns an error for an unrecognized event and it will be necessary to add a mechanism of handling this error to not crush whole execution but only log information about unsuccessful try of registering for TABLE_CHANGE event

**Comment by avelanarius** [comment]
> > Another option is to return an ERROR for the REGISTER message. Errors aren't documented for REGISTER, so we have to check if Cassandra returns an error for an unrecognized event, or if it swallows it silently.
> 
> I checked that and it depends on the driver, python-driver doesn't returns error, everything works fine, but gocql for example returns an error for an unrecognized event and it will be necessary to add a mechanism of handling this error to not crush whole execution but only log information about unsuccessful try of registering for TABLE_CHANGE event

You should also check the server-side behavior (what type of error Cassandra returns).

I think for now the behavior of the driver trying to register to `TABLET_CHANGE` event (and properly handling if the server doesn't support it) is the good way forward. For now, on the driver-side this will be only occur if the user enabled experimental tablet driver support and presumably the user will enable it if they really have a tablet Scylla cluster (but the driver should continue working even if that's not a case).

We can always add another discovery mechanism in the followup PRs.

A thing to consider: in `REGISTER` you can specify multiple events to register to. If one of them is incorrect, will/should all other or none register?

**Comment by avikivity** [comment]
The problem isn't with the driver. An unmodified driver won't ask for the event and won't need to handle any errors. A modified driver will ask for the event, but we'll also modify how it reacts to a server that doesn't support it. We need to check the existing servers (pre-tablet ScyllaDB and Cassandra). How do they react to an non-supported event?

btw, if those servers just ignore the unsupported event, that's fine too.

**Comment by sylwiaszunejko** [comment]
> The problem isn't with the driver. An unmodified driver won't ask for the event and won't need to handle any errors. A modified driver will ask for the event, but we'll also modify how it reacts to a server that doesn't support it. We need to check the existing servers (pre-tablet ScyllaDB and Cassandra). How do they react to an non-supported event?
> 
> btw, if those servers just ignore the unsupported event, that's fine too.

Both pre-tablet ScyllaDB and Cassandra return error ("Invalid value 'TABLET_CHANGE' for Event.Type"/"Invalid value 'TABLET_CHANGE' for Type"). 
Here is the place in code responsible for that:
https://github.com/scylladb/scylladb/blob/4f7e83a4d0758e4f4386ce6b128e41ca14137451/transport/server.cc#L187
https://github.com/apache/cassandra/blob/d24baecac986e90bbbb5b6281a256f41c0a27d3f/src/java/org/apache/cassandra/transport/CBUtil.java#L292

**Comment by avikivity** [comment]
Ok. So we can ask for the event and ignore it if it's not there. And in any case we shouldn't ask for it if the tablets table isn't there.

File: docs/dev/protocol-extensions.md [hunk 2]
```diff
@@ -180,3 +180,7 @@ The string map in the SUPPORTED response will contain the following parameters:
 
   - `ERROR_CODE`: a 32-bit signed decimal integer which Scylla
     will use as the error code for the rate limit exception.
+
+## Additional event
+
+When tablet table changes there is an event notification sent (`cql_transport::event::event_type::TABLET_CHANGE`). Currently it only informs that something changed without specifying details. The body of the message is empty (except the event type).
```
Comment (avelanarius) [inline docs/dev/protocol-extensions.md:186]:
```suggestion
## Tablet change event

When the tablet table changes, there is an event notification sent (`cql_transport::event::event_type::TABLET_CHANGE`). Currently it only informs that something changed without specifying details. The body of the message is empty (except the event type).
```

**Comment by avikivity** [comment]
> LGTM, apart from a small change in the docs.
> 
> @avikivity should we go ahead with merging this? My plan was to first get this PR merged, then merge (experimental) Python & gocql implementations (which will use the event) - helpful for scylla-bench and test.py, after that continue with other drivers and improving the event to contain which tablet has changed. The information that the tablets are enabled (for example in SUPPORTED message) can be done in a follow-up PR.

We need to at least have a plan for differential updates.

Maybe we should have only a differential update.

Q: what happens if a driver isn't connected to the Raft group 0 leader?

What happens if a group0 leader commits a change, then crashes before sending the event?

File: docs/dev/protocol-extensions.md (hunk 2 above)
Comment (sylwiaszunejko) [reply docs/dev/protocol-extensions.md]: Fixed

**Comment by tgrabiec** [comment]
> Q: what happens if a driver isn't connected to the Raft group 0 leader?

@avikivity on_update_tablet_metadata() is called by the state machine, so every follower will call it when it catches up with the leader. Currently, we call this only in the schema merge path. The PR needs to also extend the topology change path (starts in `group0_state_machine::merge_and_apply()` for `topology_change` command) so that we call this on tablet migration.

**Comment by tgrabiec** [comment]
The easiest way would be to call the listener in `storage_service::topology_state_load()` where we call read_tablet_metadata(). But the existing handler `storage_service::on_update_tablet_metadata()` needs to be dealt-with. It mutates token metadata, so we will have a nesting problem (deadlock). We should move the existing handler to a different callback invoked only from the schema merge path.

**Pull request closed without merging**